File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char,int> mp;
        int l=0,r=0,len=0;
        for(;r<s.size();r++){
            while(mp.find(s[r])!=mp.end()){
                char ch = s[l];
                mp[ch]--;
                if(mp[ch]==0){
                    mp.erase(ch);
                }
                l++;
            }
            mp[s[r]]++;
            len=max(len,r-l+1);
        }
        return len;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        freq = {}

        j = 0
        count = 0

        for i in range(len(s)):

            # Add current character
            freq[s[i]] = freq.get(s[i], 0) + 1

            # If duplicate, move j until duplicate is removed
            while freq[s[i]] > 1:
                freq[s[j]] -= 1
                j += 1

            # Current window length
            count = max(count, i - j + 1)

        return count

